hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/upgrade/OMPrepareRequest.java

```diff
@@ -0,0 +1,204 @@
+      RaftServerProxy server = (RaftServerProxy) omRatisServer.getServer();
+      RaftServerImpl serverImpl =
+          server.getImpl(omRatisServer.getRaftGroup().getGroupId());
+
+      takeSnapshotAndPurgeLogs(((RaftServerProxy) omRatisServer.getServer())
+          .getImpl(omRatisServer.getRaftGroup().getGroupId()));
+
+      // TODO: Create marker file with txn index.
+
+      LOG.info("OM prepared at log index {}. Returning response {}",
+          ozoneManager.getRatisSnapshotIndex(), omResponse);
+    } catch (IOException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, e));
+    } catch (InterruptedException e) {
+      response = new OMPrepareResponse(
+          createErrorOMResponse(responseBuilder, new OMException(e,
+              OMException.ResultCodes.INTERNAL_ERROR)));
+    }
+
+    return response;
+  }
+
+  private static void waitForDoubleBufferFlush(
+      OzoneManager ozoneManager, long txnLogIndex)
+      throws InterruptedException, IOException {
+
+    long endTime = System.currentTimeMillis() +
+        DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis();
+    boolean success = false;
+
+    while (!success && System.currentTimeMillis() < endTime) {
+      // If no transactions have been persisted to the DB, transaction info
+      // will be null, not zero, causing a null pointer exception within
+      // ozoneManager#getRatisSnaphotIndex.
+      // Get the transaction directly instead.
+      OMTransactionInfo txnInfo = ozoneManager.getMetadataManager()
+          .getTransactionInfoTable().get(TRANSACTION_INFO_KEY);
+      if (txnInfo == null) {
+        success = (txnLogIndex == 0);
+      } else {
+        success = (txnInfo.getTransactionIndex() == txnLogIndex);
+      }
+
+      Thread.sleep(DOUBLE_BUFFER_FLUSH_CHECK_INTERVAL.toMillis());
+    }
+
+    // If the timeout waiting for all transactions to reach the state machine
+    // is exceeded, the exception is propagated, resulting in an error response
+    // to the client. They can retry the prepare request.
+    if (!success) {
+      throw new IOException(String.format("After waiting for %d seconds, " +
+              "State Machine has not applied  all the transactions.",
+          DOUBLE_BUFFER_FLUSH_TIMEOUT.toMillis() * 1000));
+    }
+  }
+
+  /**
+   * Take a snapshot of the state machine at the last index, and purge ALL logs.
+   * @param impl RaftServerImpl instance
+   * @throws IOException on Error.
+   */
+  public static long takeSnapshotAndPurgeLogs(RaftServerImpl impl)
+      throws IOException {
+
+    StateMachine stateMachine = impl.getStateMachine();
+    long snapshotIndex = stateMachine.takeSnapshot();
+
+    // If the snapshot indices from Ratis and the state machine do not match,
+    // the exception is propagated, resulting in an error response to the
+    // client. They can retry the prepare request.
+    if (snapshotIndex != stateMachine.getLastAppliedTermIndex().getIndex()) {
```

Comment:
I believe there is a chance of snapshotIndex != lastAppliedTermIndex.

Because once doubleBuffer flush completes, it calls updateLastAppliedIndex from DoubleBuffer flush thread, to update lastAppliedIndex, but if doubleBuffer flush thread has not completed updateLastAppliedIndex updating, then this might not be equal.

There is very little chance to happen, as there is takeSnapshot which flush to DB, as updateLastAppliedIndex is inmemory map update.

So, to be safer side check lastAppliedTermIndex is the same as the TransactionInfo table lastAppliedIndex in the waitForDoubleBufferFlush method.

Reply:
The comment I put above this check is actually incorrect. This check is redundant since it is checking Ratis's in memory snapshot index against itself. It does not touch the transaction info table in the DB. I will remove this check.

> check lastAppliedTermIndex is the same as the TransactionInfo table lastAppliedIndex in the waitForDoubleBufferFlush method

I believe the check we have in the waitForDoubleBufferFlush method provides the same guarantee. Let me know if there is any added benefit to putting this check there as well.

Reply:
I understand the issue now, sorry for the confusion about the original code. It looks like this did surface on the most recent CI integration test run. I believe we should be waiting for both the lastAppliedIndex (Ratis's view) and the transaction index in the DB to update to the prepare request's index. There is no guarantee whether one or the other is updated first.